#Statistics plots
#Loading data tables...
setwd("~/PLINK/analyses/originals/AE_AEGS2_AffyAxiomGWCEU1/QC")
hwe <- read.table("aegs2_axiomgt1_b36_QCHARDY.hwe", header = T)
imiss <- read.table("aegs2_axiomgt1_b36_QCMISSING.imiss", header = T)
lmiss <- read.table("aegs2_axiomgt1_b36_QCMISSING.lmiss", header = T)
maf <- read.table("aegs2_axiomgt1_b36_QCFREQ.frq", header = T)
hwe <- read.table("aegs2_axiomgt1_b37_QCHARDY.hwe", header = T)
imiss <- read.table("aegs2_axiomgt1_b37_QCMISSING.imiss", header = T)
lmiss <- read.table("aegs2_axiomgt1_b37_QCMISSING.lmiss", header = T)
maf <- read.table("aegs2_axiomgt1_b37_QCFREQ.frq", header = T)

#Statistics plots after to sample QC
#Hardy-Weinberg equilibrium
hwe.all <- subset(hwe, hwe$TEST == "ALL")
#Graphs without y-axis limit
pdf(file="aegs2_axiomgt1_b36_QCHARDY.ALL.pdf")
hist(hwe.all$P, breaks=100, main="Hardy-Weinberg Equilibrium", xlab="P-value", ylab="# SNPs", col="steelblue", cex.axis="0.8")
dev.off()
postscript(file="aegs2_axiomgt1_b36_QCHARDY.ALL.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(hwe.all$P, breaks=100, main="Hardy-Weinberg Equilibrium", xlab="P-value", ylab="# SNPs", col="steelblue", cex.axis="0.8")
dev.off()
#Graphs with y-axis limit
pdf(file="aegs2_axiomgt1_b36_QCHARDY.ALLlim.pdf")
hist(hwe.all$P, breaks=100, main="Hardy-Weinberg Equilibrium", xlab="P-value", ylab="# SNPs", col="steelblue", cex.axis="0.8", ylim=c(1,120000))
dev.off()
postscript(file="aegs2_axiomgt1_b36_QCHARDY.ALLlim.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(hwe.all$P, breaks=100, main="Hardy-Weinberg Equilibrium", xlab="P-value", ylab="# SNPs", col="steelblue", cex.axis="0.8", ylim=c(1,120000))
dev.off()

#Sample missingness
pdf(file="aegs2_axiomgt1_b36_QCMISSING.IMISS.pdf")
hist(imiss$F_MISS, breaks=50, main="Per sample genotype call rate",pch=20, xlab="Proportion missing genotypes per sample", ylab="# Samples", xlim=c(0,0.04), ylim=c(0,100), col="steelblue", cex.axis="0.8", axes=F)
axis(2,at=c(0,10,20,30,40,50,60,70,80,90,100),tick=T)
axis(1,at=c(0,0.01,0.02,0.03,0.04),labels=c(0,0.01,0.02,0.03,0.04))
abline(v=0.03, lty=2, col="red")
dev.off()
postscript(file="aegs2_axiomgt1_b36_QCMISSING.IMISS.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(imiss$F_MISS, breaks=50, main="Per sample genotype call rate",pch=20, xlab="Proportion missing genotypes per sample", ylab="# Samples", xlim=c(0,0.04), ylim=c(0,100), col="steelblue", cex.axis="0.8", axes=F)
axis(2,at=c(0,10,20,30,40,50,60,70,80,90,100),tick=T)
axis(1,at=c(0,0.01,0.02,0.03,0.04),labels=c(0,0.01,0.02,0.03,0.04))
abline(v=0.03, lty=2, col="red")
dev.off()

#SNP missingness
pdf(file="aegs2_axiomgt1_b36_QCMISSING.LMISS.pdf")
hist(lmiss$F_MISS, breaks=500, main="Per SNP call rate", xlab="Proportion missing genotypes per SNP", ylab="# SNPs", ylim=c(0,350000), col="steelblue", cex.axis="0.8")
abline(v=0.03, lty=2, col="red")
dev.off()
postscript(file="aegs2_axiomgt1_b36_QCMISSING.LMISS.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(lmiss$F_MISS, breaks=500, main="Per SNP call rate", xlab="Proportion missing genotypes per SNP", ylab="# SNPs", ylim=c(0,350000), col="steelblue", cex.axis="0.8")
abline(v=0.03, lty=2, col="red")
dev.off()
pdf(file="aegs2_axiomgt1_b36_QCMISSING.LMISSlim.pdf")
hist(lmiss$F_MISS, breaks=500, main="Per SNP call rate", xlab="Proportion missing genotypes per SNP", ylab="# SNPs", xlim=c(0,0.2), ylim=c(0,350000), col="steelblue", cex.axis="0.8")
abline(v=0.03, lty=2, col="red")
dev.off()
postscript(file="aegs2_axiomgt1_b36_QCMISSING.LMISSlim.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(lmiss$F_MISS, breaks=500, main="Per SNP call rate", xlab="Proportion missing genotypes per SNP", ylab="# SNPs", xlim=c(0,0.2), ylim=c(0,350000), col="steelblue", cex.axis="0.8")
abline(v=0.03, lty=2, col="red")
dev.off()

#Minor allele frequency
pdf(file="aegs2_axiomgt1_b36_QCFREQ.MAF.pdf")
hist(maf$MAF, breaks=150, main="Minor allele frequency", xlab="Minor allele frequency", ylab="# SNPs", ylim=c(0,50000), col="steelblue", cex.axis="0.8")
abline(v=0.01, lty=2, col="red")
dev.off()
postscript(file="aegs2_axiomgt1_b36_QCFREQ.MAF.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(maf$MAF, breaks=150, main="Minor allele frequency", xlab="Minor allele frequency", ylab="# SNPs", ylim=c(0,50000), col="steelblue", cex.axis="0.8")
abline(v=0.01, lty=2, col="red")
dev.off()
pdf(file="aegs2_axiomgt1_b36_QCFREQ.MAFlim.pdf")
hist(maf$MAF, breaks=1000, main="Minor allele frequency", xlab="Minor allele frequency", ylab="# SNPs", xlim=c(0,0.01), ylim=c(0,50000), col="steelblue", cex.axis="0.8")
abline(v=0.01, lty=2, col="red")
dev.off()
postscript(file="aegs2_axiomgt1_b36_QCFREQ.MAFlim.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(maf$MAF, breaks=1000, main="Minor allele frequency", xlab="Minor allele frequency", ylab="# SNPs", xlim=c(0,0.01), ylim=c(0,50000), col="steelblue", cex.axis="0.8")
abline(v=0.01, lty=2, col="red")
dev.off()

#Statistics plots after to sample QC
#Hardy-Weinberg equilibrium
hwe.all <- subset(hwe, hwe$TEST == "ALL")
#Graphs without y-axis limit
pdf(file="aegs2_axiomgt1_b37_QCHARDY.ALL.pdf")
hist(hwe.all$P, breaks=100, main="Hardy-Weinberg Equilibrium", xlab="P-value", ylab="# SNPs", col="steelblue", cex.axis="0.8")
dev.off()
postscript(file="aegs2_axiomgt1_b37_QCHARDY.ALL.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(hwe.all$P, breaks=100, main="Hardy-Weinberg Equilibrium", xlab="P-value", ylab="# SNPs", col="steelblue", cex.axis="0.8")
dev.off()
#Graphs with y-axis limit
pdf(file="aegs2_axiomgt1_b37_QCHARDY.ALLlim.pdf")
hist(hwe.all$P, breaks=100, main="Hardy-Weinberg Equilibrium", xlab="P-value", ylab="# SNPs", col="steelblue", cex.axis="0.8", ylim=c(1,120000))
dev.off()
postscript(file="aegs2_axiomgt1_b37_QCHARDY.ALLlim.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(hwe.all$P, breaks=100, main="Hardy-Weinberg Equilibrium", xlab="P-value", ylab="# SNPs", col="steelblue", cex.axis="0.8", ylim=c(1,120000))
dev.off()

#Sample missingness
pdf(file="aegs2_axiomgt1_b37_QCMISSING.IMISS.pdf")
hist(imiss$F_MISS, breaks=50, main="Per sample genotype call rate",pch=20, xlab="Proportion missing genotypes per sample", ylab="# Samples", xlim=c(0,0.04), ylim=c(0,100), col="steelblue", cex.axis="0.8", axes=F)
axis(2,at=c(0,10,20,30,40,50,60,70,80,90,100),tick=T)
axis(1,at=c(0,0.01,0.02,0.03,0.04),labels=c(0,0.01,0.02,0.03,0.04))
abline(v=0.03, lty=2, col="red")
dev.off()
postscript(file="aegs2_axiomgt1_b37_QCMISSING.IMISS.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(imiss$F_MISS, breaks=50, main="Per sample genotype call rate",pch=20, xlab="Proportion missing genotypes per sample", ylab="# Samples", xlim=c(0,0.04), ylim=c(0,100), col="steelblue", cex.axis="0.8", axes=F)
axis(2,at=c(0,10,20,30,40,50,60,70,80,90,100),tick=T)
axis(1,at=c(0,0.01,0.02,0.03,0.04),labels=c(0,0.01,0.02,0.03,0.04))
abline(v=0.03, lty=2, col="red")
dev.off()

#SNP missingness
pdf(file="aegs2_axiomgt1_b37_QCMISSING.LMISS.pdf")
hist(lmiss$F_MISS, breaks=500, main="Per SNP call rate", xlab="Proportion missing genotypes per SNP", ylab="# SNPs", ylim=c(0,350000), col="steelblue", cex.axis="0.8")
abline(v=0.03, lty=2, col="red")
dev.off()
postscript(file="aegs2_axiomgt1_b37_QCMISSING.LMISS.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(lmiss$F_MISS, breaks=500, main="Per SNP call rate", xlab="Proportion missing genotypes per SNP", ylab="# SNPs", ylim=c(0,350000), col="steelblue", cex.axis="0.8")
abline(v=0.03, lty=2, col="red")
dev.off()
pdf(file="aegs2_axiomgt1_b37_QCMISSING.LMISSlim.pdf")
hist(lmiss$F_MISS, breaks=500, main="Per SNP call rate", xlab="Proportion missing genotypes per SNP", ylab="# SNPs", xlim=c(0,0.2), ylim=c(0,350000), col="steelblue", cex.axis="0.8")
abline(v=0.03, lty=2, col="red")
dev.off()
postscript(file="aegs2_axiomgt1_b37_QCMISSING.LMISSlim.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(lmiss$F_MISS, breaks=500, main="Per SNP call rate", xlab="Proportion missing genotypes per SNP", ylab="# SNPs", xlim=c(0,0.2), ylim=c(0,350000), col="steelblue", cex.axis="0.8")
abline(v=0.03, lty=2, col="red")
dev.off()

#Minor allele frequency
pdf(file="aegs2_axiomgt1_b37_QCFREQ.MAF.pdf")
hist(maf$MAF, breaks=150, main="Minor allele frequency", xlab="Minor allele frequency", ylab="# SNPs", ylim=c(0,50000), col="steelblue", cex.axis="0.8")
abline(v=0.01, lty=2, col="red")
dev.off()
postscript(file="aegs2_axiomgt1_b37_QCFREQ.MAF.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(maf$MAF, breaks=150, main="Minor allele frequency", xlab="Minor allele frequency", ylab="# SNPs", ylim=c(0,50000), col="steelblue", cex.axis="0.8")
abline(v=0.01, lty=2, col="red")
dev.off()
pdf(file="aegs2_axiomgt1_b37_QCFREQ.MAFlim.pdf")
hist(maf$MAF, breaks=1000, main="Minor allele frequency", xlab="Minor allele frequency", ylab="# SNPs", xlim=c(0,0.01), ylim=c(0,50000), col="steelblue", cex.axis="0.8")
abline(v=0.01, lty=2, col="red")
dev.off()
postscript(file="aegs2_axiomgt1_b37_QCFREQ.MAFlim.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(maf$MAF, breaks=1000, main="Minor allele frequency", xlab="Minor allele frequency", ylab="# SNPs", xlim=c(0,0.01), ylim=c(0,50000), col="steelblue", cex.axis="0.8")
abline(v=0.01, lty=2, col="red")
dev.off()


