#Statistics plots
#Loading data tables...
setwd("~/PLINK/analyses/originals/AE_AEGS2_AffyAxiomGWCEU1/QC")
hwe <- read.table("aegs2_axiomgt1_updatemap_HARDY.hwe", header = T)
imiss <- read.table("aegs2_axiomgt1_updatemap_MISSING.imiss", header = T)
lmiss <- read.table("aegs2_axiomgt1_updatemap_MISSING.lmiss", header = T)
maf <- read.table("aegs2_axiomgt1_updatemap_FREQ.frq", header = T)
stdiffmiss <- read.table("aegs2_axiomgt1_updatemap_SAMPLETYPE_MISSING.missing", header = T)
#imisshet <- read.table("aegs2_axiomgt1_updatemap_HET.het", header = T)

#Statistics plots prior to sample QC
#Hardy-Weinberg equilibrium
hwe.all <- subset(hwe, hwe$TEST == "ALL")
#Graphs without y-axis limit
pdf(file="aegs2_axiomgt1_updatemapHARDY.ALL.pdf")
hist(hwe.all$P, breaks=100, main="Hardy-Weinberg Equilibrium", xlab="P-value", ylab="# SNPs", col="steelblue", cex.axis="0.8")
dev.off()
postscript(file="aegs2_axiomgt1_updatemapHARDY.ALL.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(hwe.all$P, breaks=100, main="Hardy-Weinberg Equilibrium", xlab="P-value", ylab="# SNPs", col="steelblue", cex.axis="0.8")
dev.off()
#Graphs with y-axis limit
pdf(file="aegs2_axiomgt1_updatemapHARDY.ALLlim.pdf")
hist(hwe.all$P, breaks=100, main="Hardy-Weinberg Equilibrium", xlab="P-value", ylab="# SNPs", col="steelblue", cex.axis="0.8", ylim=c(1,120000))
dev.off()
postscript(file="aegs2_axiomgt1_updatemapHARDY.ALLlim.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(hwe.all$P, breaks=100, main="Hardy-Weinberg Equilibrium", xlab="P-value", ylab="# SNPs", col="steelblue", cex.axis="0.8", ylim=c(1,120000))
dev.off()

#Sample missingness
pdf(file="aegs2_axiomgt1_updatemapMISSING.IMISS.pdf")
hist(imiss$F_MISS, breaks=50, main="Per sample genotype call rate",pch=20, xlab="Proportion missing genotypes per sample", ylab="# Samples", xlim=c(0,0.04), ylim=c(0,100), col="steelblue", cex.axis="0.8", axes=F)
axis(2,at=c(0,10,20,30,40,50,60,70,80,90,100),tick=T)
axis(1,at=c(0,0.01,0.02,0.03,0.04),labels=c(0,0.01,0.02,0.03,0.04))
abline(v=0.03, lty=2, col="red")
dev.off()
postscript(file="aegs2_axiomgt1_updatemapMISSING.IMISS.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(imiss$F_MISS, breaks=50, main="Per sample genotype call rate",pch=20, xlab="Proportion missing genotypes per sample", ylab="# Samples", xlim=c(0,0.04), ylim=c(0,100), col="steelblue", cex.axis="0.8", axes=F)
axis(2,at=c(0,10,20,30,40,50,60,70,80,90,100),tick=T)
axis(1,at=c(0,0.01,0.02,0.03,0.04),labels=c(0,0.01,0.02,0.03,0.04))
abline(v=0.03, lty=2, col="red")
dev.off()

#SNP missingness
pdf(file="aegs2_axiomgt1_updatemapMISSING.LMISS.pdf")
hist(lmiss$F_MISS, breaks=500, main="Per SNP call rate", xlab="Proportion missing genotypes per SNP", ylab="# SNPs", ylim=c(0,350000), col="steelblue", cex.axis="0.8")
abline(v=0.03, lty=2, col="red")
dev.off()
postscript(file="aegs2_axiomgt1_updatemapMISSING.LMISS.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(lmiss$F_MISS, breaks=500, main="Per SNP call rate", xlab="Proportion missing genotypes per SNP", ylab="# SNPs", ylim=c(0,350000), col="steelblue", cex.axis="0.8")
abline(v=0.03, lty=2, col="red")
dev.off()
pdf(file="aegs2_axiomgt1_updatemapMISSING.LMISSlim.pdf")
hist(lmiss$F_MISS, breaks=500, main="Per SNP call rate", xlab="Proportion missing genotypes per SNP", ylab="# SNPs", xlim=c(0,0.2), ylim=c(0,350000), col="steelblue", cex.axis="0.8")
abline(v=0.03, lty=2, col="red")
dev.off()
postscript(file="aegs2_axiomgt1_updatemapMISSING.LMISSlim.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(lmiss$F_MISS, breaks=500, main="Per SNP call rate", xlab="Proportion missing genotypes per SNP", ylab="# SNPs", xlim=c(0,0.2), ylim=c(0,350000), col="steelblue", cex.axis="0.8")
abline(v=0.03, lty=2, col="red")
dev.off()

#Minor allele frequency
pdf(file="aegs2_axiomgt1_updatemapFREQ.MAF.pdf")
hist(maf$MAF, breaks=150, main="Minor allele frequency", xlab="Minor allele frequency", ylab="# SNPs", ylim=c(0,50000), col="steelblue", cex.axis="0.8")
abline(v=0.01, lty=2, col="red")
dev.off()
postscript(file="aegs2_axiomgt1_updatemapFREQ.MAF.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(maf$MAF, breaks=150, main="Minor allele frequency", xlab="Minor allele frequency", ylab="# SNPs", ylim=c(0,50000), col="steelblue", cex.axis="0.8")
abline(v=0.01, lty=2, col="red")
dev.off()
pdf(file="aegs2_axiomgt1_updatemapFREQ.MAFlim.pdf")
hist(maf$MAF, breaks=1000, main="Minor allele frequency", xlab="Minor allele frequency", ylab="# SNPs", xlim=c(0,0.01), ylim=c(0,50000), col="steelblue", cex.axis="0.8")
abline(v=0.01, lty=2, col="red")
dev.off()
postscript(file="aegs2_axiomgt1_updatemapFREQ.MAFlim.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(maf$MAF, breaks=1000, main="Minor allele frequency", xlab="Minor allele frequency", ylab="# SNPs", xlim=c(0,0.01), ylim=c(0,50000), col="steelblue", cex.axis="0.8")
abline(v=0.01, lty=2, col="red")
dev.off()

#Differential missingness
#Based on sample type, EDTA Blood vs. Plaque/Tissue derived DNA
pdf(file="aegs2_axiomgt1_updatemap_SAMPLETYPE_MISSING.pdf")
hist(stdiffmiss$P, breaks=100, main="Differential proportion missing genotypes per sample type", xlab="P-value", ylab="# SNPs", col="steelblue", cex.axis="0.8")
dev.off()
postscript(file="aegs2_axiomgt1_updatemap_SAMPLETYPE_MISSING.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(stdiffmiss$P, breaks=100, main="Differential proportion missing genotypes per sample type", xlab="P-value", ylab="# SNPs", col="steelblue", cex.axis="0.8")
dev.off()
pdf(file="aegs2_axiomgt1_updatemap_SAMPLETYPE_MISSINGlim.pdf")
hist(stdiffmiss$P, breaks=100000, main="Differential proportion missing genotypes per sample type", xlab="P-value", ylab="# SNPs", xlim=c(0,0.0001), ylim=c(0,10000), col="steelblue", cex.axis="0.8")
dev.off()
postscript(file="aegs2_axiomgt1_updatemap_SAMPLETYPE_MISSINGlim.eps",width=10, height=10,paper="special",horizontal=FALSE)
hist(stdiffmiss$P, breaks=100000, main="Differential proportion missing genotypes per sample type", xlab="P-value", ylab="# SNPs", xlim=c(0,0.0001), ylim=c(0,10000), col="steelblue", cex.axis="0.8")
dev.off()

#Heterozygosity rate per proportion missing SNP
imiss$logF_MISS = log10(imiss[,6])
het=read.table("aegs2_axiomgt1_updatemap_HET.het",header=T)
het$meanHet = (het$N.NM. - het$O.HOM.)/het$N.NM.
library("geneplotter")
colors  <- densCols(imiss$logF_MISS,het$meanHet)
pdf(file="aegs2_axiomgt1_updatemap_iMISS_vs_HET.pdf")
plot(imiss$logF_MISS,het$meanHet, col=colors, xlim=c(-3,0),ylim=c(0,0.5),pch=20, xlab="Proportion of missing genotypes", ylab="Heterozygosity rate",axes=F)
axis(2,at=c(0,0.05,0.10,0.15,0.2,0.25,0.3,0.35,0.4,0.45,0.5),tick=T)
axis(1,at=c(-3,-2,-1,0),labels=c(0.001,0.01,0.1,1))
abline(h=mean(het$meanHet)-(2*sd(het$meanHet)),col="RED",lty=2)
abline(h=mean(het$meanHet)+(2*sd(het$meanHet)),col="RED",lty=2)
abline(v=-1.522879, col="RED", lty=2)
dev.off()
postscript(file="aegs2_axiomgt1_updatemap_iMISS_vs_HET.eps", paper="special", width=10, height=10,horizontal=FALSE)
plot(imiss$logF_MISS,het$meanHet, col=colors, xlim=c(-3,0),ylim=c(0,0.5),pch=20, xlab="Proportion of missing genotypes", ylab="Heterozygosity rate",axes=F)
axis(2,at=c(0,0.05,0.10,0.15,0.2,0.25,0.3,0.35,0.4,0.45,0.5),tick=T)
axis(1,at=c(-3,-2,-1,0),labels=c(0.001,0.01,0.1,1))
abline(h=mean(het$meanHet)-(2*sd(het$meanHet)),col="RED",lty=2)
abline(h=mean(het$meanHet)+(2*sd(het$meanHet)),col="RED",lty=2)
abline(v=-1.522879, col="RED", lty=2)
dev.off()

